test/unit/TestContext.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Swarm unit tests for contexts
+module TestContext where
+
+import Data.Map qualified as M
+import Swarm.Language.Context
+import Swarm.Util (showT)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (Assertion, assertBool, assertEqual, testCase)
+
+testContext :: TestTree
+testContext =
+  testGroup
+    "Contexts"
+    [ testGroup
+        "Context equality"
+        [ testCase "idempotence 1" $ ctxsEqual ctx1 (ctx1 <> ctx1)
+        , testCase "idempotence 2" $ ctxsEqual ctx2 (ctx2 <> ctx2)
+        , testCase "deletion" $ ctxsEqual ctx1 (delete "z" ctx2)
+        , testCase "empty/delete" $ ctxsEqual empty (delete "x" ctx1)
+        , testCase "fromMap" $ ctxsEqual ctx2 (fromMap (M.fromList [("x", 3), ("z", 6)]))
+        , testCase "right bias" $ ctxsEqual ctx4 (ctx2 <> ctx3)
+        , testCase "commutativity" $ ctxsEqual (ctx1 <> ctx5) (ctx5 <> ctx1)
+        ]
+    , testGroup
+        "de/rehydrate round-trip"
+        [ testCase "empty" $ serializeRoundTrip empty
+        , testCase "ctx1" $ serializeRoundTrip ctx1
+        , testCase "ctx2" $ serializeRoundTrip ctx2
+        , testCase "ctx3" $ serializeRoundTrip ctx3
+        , testCase "ctx4" $ serializeRoundTrip ctx4
+        , testCase "ctx5" $ serializeRoundTrip ctx5
+        , testCase "large" $ serializeRoundTrip bigCtx
+        , testCase "delete" $ serializeRoundTrip (delete "y" ctx4)
+        ]
+    ]
+ where
+  ctx1 = singleton "x" 3
+  ctx2 = singleton "x" 3 <> singleton "z" 6
+  ctx3 = singleton "x" 5 <> singleton "y" 7
+  ctx4 = singleton "x" 5 <> singleton "y" 7 <> singleton "z" 6
+  ctx5 = singleton "y" 10
+  bigCtx = fromMap . M.fromList $ zip (map (("x" <>) . showT) [1 :: Int ..]) [1 .. 10000]
+
+ctxsEqual :: Ctx Int -> Ctx Int -> Assertion
+ctxsEqual ctx1 ctx2 = do
```

Comment:
How about contexts not being equal?

```math
c_1 = c_2 \equiv hash c_1 = hash c_2
```

I think the equivalence tests do not prevent the trivial implementation `hash x = 0`.

Reply:
I mean, the requirement to also have structural equality in addition to hash equality would not allow `hash x = 0` to pass.  But it's a good point, and we also want to make sure that hash collisions are as rare as I think they should be.  I will add some QuickCheck tests to make sure that randomly generated contexts have hashes that are equal iff the resulting maps are equal.